ReactiveObjC/RACSignal.m

```diff
@@ -25,8 +25,56 @@
 #import "RACTuple.h"
 #import <stdatomic.h>
 
+#if DEBUG
+
+#import "RACSourceSymbolExtractor.h"
+
+@interface RACSignal ()
+
+@property (readonly, nonatomic) NSArray<NSString *> *initializationCallStackSymbols;
+
+@property (readonly, nonatomic) NSLock *initializationSourceSymbolLock;
+
+@property (nonatomic) BOOL didExtractInitializationSourceSymbol;
+
+@end
+
+#endif
+
 @implementation RACSignal
 
+#ifdef DEBUG
+
+@synthesize initializationSourceSymbol = _initializationSourceSymbol;
+
+- (instancetype)init {
+  if (self = [super init]) {
+    // Lightweight.
+    _initializationCallStackSymbols = NSThread.callStackSymbols;
```

Comment:
Why do you claim that this is a lightweight operation? You need to iterate the stack (potentially fast) and fetch symbol per address.

Reply:
I should have had a disclaimer that this is based on my observations and conclusion but this is lightweight:
First by debugging we can see that the type returned from `NSThread.callStackSymbols` is not a "regular" `NSArray` where all values are necessarily stored in advance but of type `_NSCallStackArray` i.e. it has a designated NSArray descendant type.
![Snip20210408_4](https://user-images.githubusercontent.com/15310405/114006375-1074a500-9869-11eb-9444-dcf0a9483011.png)

Second: from some measurements I did, it can be seen that only accessing the elements in the array is heavy while the call to `NSThread.callStackSymbols` is lightweight.
I decided to do a more organized benchmarking and measure this on a stack with some (relatively small) depth:
This is the average time for the `NSThread.callStackSymbols` on my iPhoneX over 100 iterations: `8.03751e-07`.
And this is the average time over that 100 iterations for accessing each one of the `callStackSymbols` array elements (one line per element):
```
(
    "0.001459755005780607",
    "5.995825631543994e-07",
    "4.620820982381702e-07",
    "3.250018926337361e-07",
    "3.624981036409736e-07",
    "3.062532050535083e-07",
    "2.754165325313806e-07",
    "4.479195922613144e-07",
    "3.620836650952697e-07",
    "2.737512113526464e-07",
    "3.94163653254509e-07",
    "3.708328586071729e-07",
    "3.175029996782541e-07",
    "3.254151670262218e-07",
    "3.275007475167513e-07",
    "6.66668638586998e-07",
    "3.774987999349833e-07"
)
```
As can easily be seen - the heavy time is done by accessing the first element in the array. Because the times for the other elements is small, this also might imply that the symbolication for all the elements is done on the first element (or that the heavy part of the computation is cached).
That suggest that indeed doing this per element and stop where reaching the first non ReactiveObjc can significantly speed up the computation

Reply:
Added to code of the benchmarking to the PR

Reply:
Measures with `dladdr`:
Average time for `backtrace` method: `3.99165e-07`
And for each element:
```
    "5.979154957458377e-07",
    "4.787516081705689e-07",
    "4.533358151093125e-07",
    "4.687515320256353e-07",
    "4.545855335891247e-07",
    "4.612485645338893e-07",
    "4.42083110101521e-07",
    "4.533323226496577e-07",
    "0.0005292091658338905",
    "0.0002010483329650015",
    "0.0002008183335419744",
    "0.0001114299998153001",
    "0.0002010270836763084",
    "0.0002008824987569824",
    "0.0002045900013763458",
    "8.724996587261558e-07",
    "6.898754509165883e-06"
```
Indeed much better.
\+ `dladdr` populates a struct called `Dl_info` with the symbol components already decomposed - so no need in all the regex stuff

Reply:
Cool. By looking at Foundations, your observations are right:
- `NSThread.callStackSymbols` is calling `backtrace` under the hood.
- `_NSCallStackArray` lazily calls `backtrace_symbols` when accessing the first element if it was initialized with a truthy `symbols` argument.